### PR TITLE
Adding dockerfile component

### PR DESCRIPTION
**Pull request by ranakan19**
### What does this PR do?
Adds dockerfile component for devfile 2.1.0

### What issues does this PR fix or reference?
In reference to https://github.com/devfile/parser/issues/26

### Is your PR tested? Consider putting some instruction how to test your changes
Not applicable.

### CODE REVIEW

File: pkg/apis/workspaces/v1alpha1/dockerfileComponent.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+	// Mandatory path to source code
+	Source *Source `json:"source"`
+
+	// Mandatory path to dockerfile
+	DockerfileLocation string `json:"dockerfileLocation"`
+
+	// Mandatory destination to registry to push built image
+	Destination string `json:"destination,omitempty"`
+}
```
Comment (wtam2018) [inline pkg/apis/workspaces/v1alpha1/dockerfileComponent.go:16]:
```suggestion
	
	// Mandatory whether rootless/unprivileged builder container is required
	Rootless bool `json:"rootless,omitempty"`
}
```

**Comment by kadel** [comment]
This is targeted to 2.1.0 but we have only a single version of types. How is this versioning going to work?

**Comment by ranakan19** [comment]
> This is targeted to 2.1.0 but we have only a single version of types. How is this versioning going to work?

I'm not sure i understand your question, do you mean the schemaVersion? In the devfile library, devfile 2.0.0 and 2.1.0 are separate schema, but i figured that the kubernetes-api can have all the common structs to be used.

**Comment by kadel** [comment]
> > This is targeted to 2.1.0 but we have only a single version of types. How is this versioning going to work?
> 
> @kadel I'm not sure i understand your question, do you mean the schemaVersion? In the devfile library, devfile 2.0.0 and 2.1.0 are separate schema, but i figured that the kubernetes-api can have all the common structs to be used.

Yes, I'm wondering how is this going to work with `schemaVersion`.

For example, If I have a devfile that has `schemaVersion: 2.0.0` with Dockerfile component.
Would that get parsed or not?

**Comment by kadel** [comment]
btw @ranakan19, you need to run `./docker-run.sh ./build.sh` to generate updated Json Schema and other autogenerated files. Autogenerated files should be included in your PR as a separate commit.

**Comment by davidfestal** [comment]
> > > This is targeted to 2.1.0 but we have only a single version of types. How is this versioning going to work?
> > 
> > 
> > @kadel I'm not sure i understand your question, do you mean the schemaVersion? In the devfile library, devfile 2.0.0 and 2.1.0 are separate schema, but i figured that the kubernetes-api can have all the common structs to be used.
> 
> Yes, I'm wondering how is this going to work with `schemaVersion`.
> 
> For example, If I have a devfile that has `schemaVersion: 2.0.0` with Dockerfile component.
> Would that get parsed or not?

I assume that at some point we would have to keep several versions of the K8S apis (like the v1, v2alpha1, etc, ...), and possibly maintain a mapping between a devfile schema version (semantic-versioning compatible) and the corresponding K8S API version.

That's something that is done for K8S itself for example. But we should obviously discuss how this could be achieved as simply as possible in our case, knowing that we want to avoid in upcoming devfile schema versions, changes that would introduce incompatible changes.

**Comment by davidfestal** [comment]
I created PR https://github.com/ranakan19/kubernetes-api/pull/2 to fix some small errors, and add required code for correct union management.

**Comment by davidfestal** [comment]
I have 2 other questions:

1. In this code block:
```go
//Source within dockerfile component
type Source struct {
	// path to local source directory folder
	SourceDir string `json:"sourceDir,omitempty"`

	// path to source repository hosted locally or on cloud
	Location string `json:"location,omitempty"`
}
```
Are the 2 fields mutually exclusive and can they *both* have a value defined.
If they are mutually exclusive, it could be defined as a union, so that the schema itself would ensure this union-member oneness.

2. Does overriding the dockerimage component of a parent makes sense ?
For now the dockerimage is not added into the overrides. And if yes, does it also make sense to override a dockerfile component that would be encapsulated in a plugin definition ?

**Comment by wtam2018** [comment]
> Are the 2 fields mutually exclusive and can they _both_ have a value defined.
They are mutually exclusive but `SourceDir` does not seem to make sense to be in there.   We can have `SourceLocation` string instead of `Source` struct.

**Comment by ranakan19** [comment]
> I have 2 other questions:
> 
> 1. In this code block:
> 
> ```go
> //Source within dockerfile component
> type Source struct {
> 	// path to local source directory folder
> 	SourceDir string `json:"sourceDir,omitempty"`
> 
> 	// path to source repository hosted locally or on cloud
> 	Location string `json:"location,omitempty"`
> }
> ```
> 
> Are the 2 fields mutually exclusive and can they _both_ have a value defined.
> If they are mutually exclusive, it could be defined as a union, so that the schema itself would ensure this union-member oneness.
> 
> 1. Does overriding the dockerimage component of a parent makes sense ?
>    For now the dockerimage is not added into the overrides. And if yes, does it also make sense to override a dockerfile component that would be encapsulated in a plugin definition ?

for 1, Yes the fields are mutually exclusive. I'm not sure about 2, we should discuss it in the meeting.

**Review by wtam2018 — COMMENTED** [review]
This PR is superseded by https://github.com/devfile/api/pull/127.   Please close this one.

**Comment by sleshchenko** [comment]
Closing in favor of https://github.com/devfile/api/pull/127

**Pull request closed without merging**